Split RzBin PE parser into pieces

PeiweiHu opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Split gigantic C file into multiple, like it was done for ELF before.

**Test plan**

CI is green. 

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/2614
<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

PeiweiHu committed:
Split RzBin PE parser into smaller pieces

XVilka commented:
@PeiweiHu please rebase to resolve the conflicts. Note also the newer parts that are necessary for the .NET parsing.

XVilka commented on librz/bin/format/pe/pe_debug.c:
```diff
@@ -0,0 +1,187 @@
+#include "pe.h"
```
All files should have SPDX headers.

PeiweiHu commented:
> @PeiweiHu please rebase to resolve the conflicts. Note also the newer parts that are necessary for the .NET parsing.

Thanks for your review, they are rebased now.

wargio commented on librz/bin/meson.build:
```diff
@@ -205,6 +205,29 @@ rz_bin_sources = [
+  'format/pe/pe64_clr.c',
+  'format/pe/pe_debug.c',
+  'format/pe/pe64_debug.c',
+  'format/pe/pemixed.c',
   'format/pe/pemixed.c',
```
dup?

XVilka reviewed (approved):
If I understood correctly, you didn't change the code, right? Just refactored to split it?
Then, it's good to be merged. Further improvements should be done separately for easier review.

PeiweiHu commented:
> If I understood correctly, you didn't change the code, right? Just refactored to split it? Then, it's good to be merged. Further improvements should be done separately for easier review.

Yep, I just split the existing code.

PeiweiHu commented:
> If I understood correctly, you didn't change the code, right? Just refactored to split it? Then, it's good to be merged. Further improvements should be done separately for easier review.

Plus, I changed the names of several APIs to expose them for refactor. For example, `bin_pe_rva_to_paddr `is changed to `PE_(bin_pe_rva_to_paddr)`. But there are no functional changes.

ret2libc commented on librz/bin/format/pe/pe.h:
```diff
@@ -170,45 +170,80 @@ struct PE_(rz_bin_pe_obj_t) {
+void *PE_(rz_bin_pe_free)(struct PE_(rz_bin_pe_obj_t) * bin);
+struct PE_(rz_bin_pe_obj_t) * PE_(rz_bin_pe_new)(const char *file, bool verbose);
+struct PE_(rz_bin_pe_obj_t) * PE_(rz_bin_pe_new_buf)(RzBuffer *buf, bool verbose);
+// unimplemented
```
If they are not implemented, can we just remove them?

ret2libc commented on librz/bin/format/pe/pe.h:
```diff
@@ -170,45 +170,80 @@ struct PE_(rz_bin_pe_obj_t) {
+// pe_tls.c
+int PE_(bin_pe_init_tls)(struct PE_(rz_bin_pe_obj_t) * bin);
+// pe.c
+#define RzBinPEObj struct PE_(rz_bin_pe_obj_t)
```
move this at the top of the file and use it everywhere instead of using `struct PE_(rz_bin_pe_obj_t)`.

ret2libc commented on librz/bin/format/pe/pe.h:
```diff
@@ -170,45 +170,80 @@ struct PE_(rz_bin_pe_obj_t) {
+int PE_(bin_pe_init_imports)(struct PE_(rz_bin_pe_obj_t) * bin);
+int PE_(read_image_import_directory)(RzBuffer *b, ut64 addr, PE_(image_import_directory) * import_dir);
+int PE_(read_image_delay_import_directory)(RzBuffer *b, ut64 addr, PE_(image_delay_import_directory) * directory);
 struct rz_bin_pe_import_t *PE_(rz_bin_pe_get_imports)(struct PE_(rz_bin_pe_obj_t) * bin); // TODO
```
remove these TODOs please if it is not clear what needs to be done

ret2libc commented on librz/bin/format/pe/pe.h:
```diff
@@ -170,45 +170,80 @@ struct PE_(rz_bin_pe_obj_t) {
+// pe_overlay.c
+int PE_(bin_pe_get_overlay)(struct PE_(rz_bin_pe_obj_t) * bin, ut64 *size);
+int PE_(bin_pe_init_overlay)(struct PE_(rz_bin_pe_obj_t) * bin);
+// pe_rsrc.c
```
also please put a newline before each file name in this file, as there are many functions and i believe having a new line will improve the detection of each "block" of functions

ret2libc commented on librz/bin/format/pe/pe.c:
```diff
@@ -396,4126 +25,64 @@ static PE_DWord bin_pe_rva_to_paddr(RzBinPEObj *bin, PE_DWord rva) {
-	record
-	paddr
+	bin->sections = PE_(rz_bin_pe_get_sections)(bin);
+	PE_(bin_pe_init_imports)
+	(bin);
+	PE_(bin_pe_init_exports)
+	(bin);
+	PE_(bin_pe_init_resource)
+	(bin);
+	PE_(bin_pe_init_security)
+	(bin);
```
fix the style of these please

ret2libc commented on librz/bin/format/pe/pe.c:
```diff
@@ -396,4126 +25,64 @@ static PE_DWord bin_pe_rva_to_paddr(RzBinPEObj *bin, PE_DWord rva) {
+	bin->big_endian = PE_(rz_bin_pe_is_big_endian)(bin);
 
-#endif
+	PE_(bin_pe_init_rich_info)
+	(bin);
+	PE_(bin_pe_init_tls)
+	(bin);
+	PE_(bin_pe_init_clr)
+	(bin);
+	PE_(bin_pe_init_overlay)
+	(bin);
+	PE_(bin_pe_parse_resource)
+	(bin);
```
also this

ret2libc commented on librz/bin/format/pe/pe.c:
```diff
@@ -4528,7 +95,8 @@ void *PE_(rz_bin_pe_free)(struct PE_(rz_bin_pe_obj_t) * bin) {
 	free(bin->import_directory);
 	free(bin->resource_directory);
-	free_security_directory(bin->security_directory);
+	PE_(free_security_directory)
+	(bin->security_directory);
```
style

ret2libc commented on librz/bin/format/pe/pe.c:
```diff
@@ -6,384 +6,13 @@
-	return winmain;
-}
 
 #define RzBinPEObj struct PE_(rz_bin_pe_obj_t)
```
not necessary anymore?

ret2libc commented on librz/bin/format/pe/pe_debug.c:
```diff
@@ -0,0 +1,192 @@
+
+#include "pe.h"
+
+struct SCV_NB10_HEADER;
```
why this (and the others)?
Isn't it enough to have the typedef below?

PeiweiHu replied:
They are not removed to avoid the compiler warnings.

ret2libc reviewed (commented):
Was there any file where you applied significant changes apart from moving functions from pe.c into specific files?

PeiweiHu commented:
> Was there any file where you applied significant changes apart from moving functions from pe.c into specific files?

Thank you for your thoughtful reviews. This pr focuses on splitting files instead of functional changes. Most of the places are refined as you suggested except one, as shown in `unresolved conversation`.

XVilka commented:
@PeiweiHu clang-format still produces error

PeiweiHu commented:
> @PeiweiHu clang-format still produces error

@XVilka The errors stem from the formats like `PE_(bin_pe_init_imports)(bin);`, which is suggested by @ret2libc. Should I change it back (`PE_(bin_pe_init_imports)\n(bin);`) to conform to clang-format?

ret2libc commented:
> > @PeiweiHu clang-format still produces error
> 
> @XVilka The errors stem from the formats like `PE_(bin_pe_init_imports)(bin);`, which is suggested by @ret2libc. Should I change it back (`PE_(bin_pe_init_imports)\n(bin);`) to conform to clang-format?

@XVilka do you have any idea why that happens? I'm not sure it happens with Elf_ macro as well

XVilka commented:
@ret2libc @PeiweiHu it was like that even before. I guess it's clang-format bug, I think it's a minor inconvenience but should not stop this very useful refactoring.
